invocation.md

```diff
@@ -0,0 +1,1170 @@
+
+```json
+// CID
+{"/": "Qmf412jQZiuVUtdgnB36FXFX7xg5V6KEbSJ4dpQuhkLyfD"}
+
+// Bytes (e.g. a signature)
+{"/": {"bytes": "s0m3Byte5"}}
+```
+
+This format help disambiguate type information in generic [DAG-JSON] tooling. However, your presentation need not be in this specific format, as long as it can be converted to and from this cleanly. As it is used for the signature format, [DAG-CBOR] is RECOMMENDED.
+
+## 1.4 Note on Schema Syntax
+
+We use [IPLD Schema] syntax extended with generics. Standard IPLD Schema can be derived by ignoring parameters enclosed in angle brackets and interpreting parameters as `Any`.
+
+Below schema is in our extended syntax
+
+```ipldsch
+type Box<T> struct {
+  value T
+}
+```
+
+Above schema compiled to standard syntax
+
+```ipldsch
+type Box struct {
+  value Box_T
+}
+type Box_T any
+```
+
+## 1.5 Signatures
+
+All payloads described in this spec MUST be signed with a [Varsig].
+
+# 2 High-Level Concepts
+
+## 2.1 Roles
+
+Task adds two new roles to UCAN: invoker and executor. The existing UCAN delegator and delegate principals MUST persist to the invocation.
+
+| UCAN Field | Delegation                             | Task                            |
+| ---------- | -------------------------------------- | ------------------------------- |
+| `iss`      | Delegator: transfer authority (active) | Invoker: request task (active)  |
+| `aud`      | Delegate: gain authority (passive)     | Executor: perform task (active) |
+
+### 2.1.1 Invoker
+
+The invoker signals to the executor that a task associated with a UCAN SHOULD be performed.
+
+The invoker MUST be the UCAN delegator. Their DID MUST be authenticated in the `iss` field of the contained UCAN.
+
+### 2.1.2 Executor
+
+The executor is directed to perform some task described in the UCAN by the invoker.
+
+The executor MUST be the UCAN delegate. Their DID MUST be set the in `aud` field of the contained UCAN.
+
+## 2.2 Components
+
+```mermaid
+flowchart LR
+ U(UCAN) -- Auth --> I
+ I(Invocation) -- Output --> O(Result)
+ O -- Consume --> P(Promise)
+ O -- Attest --> R(Receipt)
+ R -- Short-circuit --> M(Memoization)
+```
+
+### 2.2.1 Invocation
+
+An [Invocation] is like a cryptographically signed function application, a request to perform some action on a resource with specific input. Invocation MAY have optional metadata that is not used to describe the meaning of the computation or effect to be run. Executor MUST produce same result regardless of the metadata.
+
+### 2.2.2 Receipt
+
+A [Receipt] describes the output of an invocation. It is referenced either by the invocation itself or an IPLD Link of the invocation.
+
+### 2.2.3 Promise
+
+A [promise] is a reference to expected result of the invocation receipt.
+
+## 2.3 IPLD Schema
+
+```ipldsch
+type Invocation<In> struct {
+  v      SemVer
+
+  iss    Principal
+  aud    Principal
+
+  with   URI
+  do     Ability
+  input  In (implicit {})
```

Comment:
if we're shortening key names like `s` for the varsig, maybe this could be `in`? fwiw I don't think that shortname is claimed in the jwt claim registry that we borrowed/were-inspired-by e.g. `nbf` and `iss` from https://www.iana.org/assignments/jwt/jwt.xhtml#claims

Reply:
also noticing that receipt has `out` not `output` so `in` may fit better aesthetically, regardless of length

Reply:
I actually wanted to use `in`, but was hesitating for the same reason field named `with` is really annoying (in JS) as you can't destructure or have variables with that name.

I don't mind renaming it to `in` though

Reply:
I have also considered `use` because then you have "with Resource do Action, use Input"